File: mobile/modules/frame-preview/android/src/main/java/com/mentra/framepreview/PreviewStats.kt
package com.mentra.framepreview

import kotlin.math.roundToInt

/** One ring's shape, produced by a single sort. */
data class PreviewDistribution(
  val p50: Double,
  val p95: Double,
  val p99: Double,
  val max: Double,
  val count: Int,
) {
  companion object {
    val EMPTY = PreviewDistribution(0.0, 0.0, 0.0, 0.0, 0)
  }
}

/**
 * Fixed-size sample ring with numeric percentiles and a running maximum.
 *
 * Numeric rather than the formatted helper in `glasses-media`: these values cross into the
 * WebView as JSON and the page charts them, so "na" in a number field would be a parse error
 * rather than a missing sample.
 *
 * The capacity is part of the claim being made. At 30 fps a 64-sample ring only describes the
 * last two seconds, so a p95 read off it after a ten-minute soak is a two-second p95 wearing a
 * ten-minute label. [maxMs] is deliberately not windowed: the worst frame of the run is the one
 * that shows up as a visible hitch, and a lapped ring would have thrown it away.
 */
class PreviewPercentileRing(private val capacity: Int = 2048) {
  private val samples = LongArray(capacity)
  private var count = 0
  private var index = 0
  private var maxNs = 0L

  @Synchronized
  fun record(valueNs: Long) {
    samples[index] = valueNs
    index = (index + 1) % capacity
    if (count < capacity) count += 1
    if (valueNs > maxNs) maxNs = valueNs
  }

  /**
   * Every percentile from one sort.
   *
   * Reading p50, p95 and p99 as three separate calls sorts the ring three times, and the
   * reporter reads seven rings every second. At 30 fps that is tens of thousands of needless
   * comparisons per second while frames are being packed — the measurement showing up inside
   * the thing being measured.
   */
  @Synchronized
  fun distribution(): PreviewDistribution {
    if (count == 0) return PreviewDistribution.EMPTY
    val sorted = samples.copyOf(count)
    sorted.sort()
    fun at(quantile: Double): Double {
      val position = ((sorted.size - 1) * quantile).roundToInt().coerceIn(0, sorted.lastIndex)
      return Math.round(sorted[position] / 1_000_000.0 * 100.0) / 100.0
    }
    return PreviewDistribution(at(0.50), at(0.95), at(0.99), maxMs(), count)
  }

  @Synchronized
  fun percentileMs(quantile: Double): Double {
    if (count == 0) return 0.0
    val sorted = samples.copyOf(count)
    sorted.sort()
    val position = ((sorted.size - 1) * quantile).roundToInt().coerceIn(0, sorted.lastIndex)
    return Math.round(sorted[position] / 1_000_000.0 * 100.0) / 100.0
  }

  @Synchronized
  fun maxMs(): Double = Math.round(maxNs / 1_000_000.0 * 100.0) / 100.0

  @Synchronized
  fun sampleCount(): Int = count

  @Synchronized
  fun reset() {
    count = 0
    index = 0
    maxNs = 0
  }
}

/**
 * Counters for one preview subscription.
 *
 * Every drop has its own counter on purpose: "delivered fps is low" is not a finding, but
 * "delivered is low because the pacer skipped, not because the consumer was busy" is.
 *
 * The timing rings are split finer than they strictly need to be for the same reason. A single
 * `sendMs` would average an enqueue that returns immediately together with a copy that has not
 * happened yet, and the two platforms would then be reporting different things under one name.
 */
class PreviewStats {
  @Volatile
  var sourceFrames = 0L
    private set

  @Volatile
  var admitted = 0L
    private set

  @Volatile
  var skippedPacing = 0L
    private set

  @Volatile
  var skippedBusy = 0L
    private set

  @Volatile
  var delivered = 0L
    private set

  @Volatile
  var payloadBytes = 0L
    private set

  @Volatile
  var ackTimeouts = 0L
    private set

  @Volatile
  var staleAcks = 0L
    private set

  @Volatile
  var packFailures = 0L
    private set

  @Volatile
  var transportErrors = 0L
    private set

  /**
   * Frames refused on the decoder thread because the worker had not finished the previous one.
   * Counted apart from [skippedBusy] so a decoder-side refusal is not confused with a
   * consumer-side one; they have different fixes.
   */
  @Volatile
  var preDispatchDrops = 0L
    private set

  /** The pack worker wanted a send buffer and both were still held by the transport. */
  @Volatile
  var slotStarved = 0L
    private set

  /** Pages that presented the current document's token. */
  @Volatile
  var handshakes = 0L
    private set

  /** Android fallback reloads because the listener was installed after navigation. Should stay 0. */
  @Volatile
  var installReloads = 0L
    private set

  /** Output tier changes applied without restarting production or the transport. */
  @Volatile
  var tierChanges = 0L
    private set

  /** Throwables caught from the tap sink on the decoder thread, accumulated across drains. */
  @Volatile
  var tapSinkExceptions = 0L
    private set

  private val packFailureCounts = LinkedHashMap<String, Long>()

  /**
   * Producing the source frame. Only the synthetic source pays this; a real decoder hands over a
   * buffer it has already made. Kept apart from [pack] so a slow test pattern cannot be read as
   * a slow pipeline.
   */
  val generate = PreviewPercentileRing()

  val pack = PreviewPercentileRing()

  /**
   * Duration of `postMessage` itself, on the UI thread. Chromium copies inside that call, so on
   * Android this is the whole send; iOS has to split it because its enqueue returns early.
   */
  val sendComplete = PreviewPercentileRing()

  val roundTrip = PreviewPercentileRing()

  /** Admission to the first instruction of the pack worker: the handoff cost, not the work. */
  val admitToPack = PreviewPercentileRing()

  /** Gap between consecutive delivered frames. Mean fps hides a 30/100 ms alternation; this does not. */
  val deliveryGap = PreviewPercentileRing()

  /** How long the UI-thread send runnable waited before it ran. The app's jank, not ours to hide. */
  val mainQueueWait = PreviewPercentileRing()

  /**
   * Start of run to first delivered frame, and to its acknowledgement. A slow cold start is
   * invisible in steady-state fps but very visible to someone opening the panel.
   */
  @Volatile
  var firstDeliveredLatencyMs = 0.0
    private set

  @Volatile
  var firstAckLatencyMs = 0.0
    private set

  // Nullable rather than a zero sentinel. A monotonic clock reading of exactly zero is unlikely
  // on a device but trivial in a test, and "not yet set" and "set to zero" have to stay distinct
  // or the first gap of every run goes unrecorded.
  private var runStartNs: Long? = null
  private var lastDeliveredAtNs: Long? = null
  private var hasFirstDelivered = false
  private var hasFirstAck = false

  private var windowStartNs: Long? = null
  private var windowSourceFrames = 0L
  private var windowDelivered = 0L
  private var windowPayloadBytes = 0L

  @Synchronized
  fun onSourceFrame() {
    sourceFrames++
    windowSourceFrames++
  }

  @Synchronized
  fun onAdmitted() {
    admitted++
  }

  @Synchronized
  fun onSkippedPacing() {
    skippedPacing++
  }

  @Synchronized
  fun onSkippedBusy() {
    skippedBusy++
  }

  @Synchronized
  fun onAckTimeout() {
    ackTimeouts++
  }

  @Synchronized
  fun onStaleAck() {
    staleAcks++
  }

  @Synchronized
  fun onPackFailure(reason: PackFailureReason) {
    packFailures++
    packFailureCounts[reason.wire] = (packFailureCounts[reason.wire] ?: 0L) + 1
  }

  /** Pack failures keyed by [PackFailureReason.wire]; only reasons that occurred are present. */
  @Synchronized
  fun packFailuresByReason(): Map<String, Long> = LinkedHashMap(packFailureCounts)

  @Synchronized
  fun onHandshake() {
    handshakes++
  }

  @Synchronized
  fun onInstallReload() {
    installReloads++
  }

  @Synchronized
  fun onTierChange() {
    tierChanges++
  }

  @Synchronized
  fun onTapSinkExceptions(count: Long) {
    tapSinkExceptions += count
  }

  @Synchronized
  fun onTransportError() {
    transportErrors++
  }

  @Synchronized
  fun onPreDispatchDrop() {
    preDispatchDrops++
  }

  @Synchronized
  fun onSlotStarved() {
    slotStarved++
  }

  /**
   * Mark the beginning of a run so the first-frame latencies have an origin.
   *
   * This also restarts the rate window. Without that, the first second after a stop/start would
   * be averaged over however long the preview sat idle and report a frame rate near zero.
   */
  @Synchronized
  fun onRunStart(nowNs: Long) {
    runStartNs = nowNs
    lastDeliveredAtNs = null
    hasFirstDelivered = false
    hasFirstAck = false
    firstDeliveredLatencyMs = 0.0
    firstAckLatencyMs = 0.0
    windowStartNs = nowNs
    windowSourceFrames = 0
    windowDelivered = 0
    windowPayloadBytes = 0
  }

  @Synchronized
  fun onDelivered(bytes: Int, nowNs: Long) {
    delivered++
    windowDelivered++
    payloadBytes += bytes
    windowPayloadBytes += bytes
    lastDeliveredAtNs?.let { deliveryGap.record(nowNs - it) }
    lastDeliveredAtNs = nowNs
    val start = runStartNs
    if (!hasFirstDelivered && start != null) {
      hasFirstDelivered = true
      firstDeliveredLatencyMs = (nowNs - start) / 1_000_000.0
    }
  }

  @Synchronized
  fun onAckAccepted(roundTripNs: Long, nowNs: Long) {
    roundTrip.record(roundTripNs)
    val start = runStartNs
    if (!hasFirstAck && start != null) {
      hasFirstAck = true
      firstAckLatencyMs = (nowNs - start) / 1_000_000.0
    }
  }

  data class Window(val sourceFps: Double, val deliveredFps: Double, val bytesPerSecond: Double)

  /** Read and restart the rate window. Zeros on the first call of a subscription. */
  @Synchronized
  fun takeWindow(nowNs: Long): Window {
    val start = windowStartNs
    windowStartNs = nowNs
    val frames = windowSourceFrames
    val deliveredInWindow = windowDelivered
    val bytes = windowPayloadBytes
    windowSourceFrames = 0
    windowDelivered = 0
    windowPayloadBytes = 0
    if (start == null || nowNs <= start) return Window(0.0, 0.0, 0.0)
    val seconds = (nowNs - start) / 1_000_000_000.0
    return Window(frames / seconds, deliveredInWindow / seconds, bytes / seconds)
  }

  @Synchronized
  fun reset(nowNs: Long) {
    sourceFrames = 0
    admitted = 0
    skippedPacing = 0
    skippedBusy = 0
    delivered = 0
    payloadBytes = 0
    ackTimeouts = 0
    staleAcks = 0
    packFailures = 0
    transportErrors = 0
    preDispatchDrops = 0
    slotStarved = 0
    handshakes = 0
    installReloads = 0
    tierChanges = 0
    tapSinkExceptions = 0
    packFailureCounts.clear()
    generate.reset()
    pack.reset()
    sendComplete.reset()
    roundTrip.reset()
    admitToPack.reset()
    deliveryGap.reset()
    mainQueueWait.reset()
    firstDeliveredLatencyMs = 0.0
    firstAckLatencyMs = 0.0
    hasFirstDelivered = false
    hasFirstAck = false
    runStartNs = nowNs
    lastDeliveredAtNs = null
    windowStartNs = nowNs
    windowSourceFrames = 0
    windowDelivered = 0
    windowPayloadBytes = 0
  }
}

File: mobile/modules/frame-preview/ios/PreviewKit/Sources/FramePreviewCore/PreviewStats.swift
import Foundation

/// One ring's shape, produced by a single sort.
public struct PreviewDistribution: Equatable, Sendable {
  public let p50: Double
  public let p95: Double
  public let p99: Double
  public let max: Double
  public let count: Int

  public static let empty = PreviewDistribution(p50: 0, p95: 0, p99: 0, max: 0, count: 0)
}

/// Fixed-size sample ring with percentiles and a running maximum.
///
/// The capacity is part of the claim being made. At 30 fps a 64-sample ring only describes the
/// last two seconds, so a p95 read off it after a ten-minute soak is a two-second p95 wearing a
/// ten-minute label. 2048 samples is a little over a minute at 30 fps, which is the shortest run
/// the experiment reports on.
///
/// `maxMs` is deliberately not windowed: the worst frame of the run is the one that shows up as a
/// visible hitch, and a ring that has already lapped would have thrown it away.
public final class PreviewPercentileRing {
  private var samples: [Int64]
  private var count = 0
  private var index = 0
  private var maxNs: Int64 = 0

  public init(capacity: Int = 2048) {
    samples = [Int64](repeating: 0, count: max(capacity, 1))
  }

  public var sampleCount: Int { count }

  public var maxMs: Double { Double(maxNs) / 1_000_000.0 }

  public func record(_ valueNs: Int64) {
    samples[index] = valueNs
    index = (index + 1) % samples.count
    if count < samples.count { count += 1 }
    if valueNs > maxNs { maxNs = valueNs }
  }

  /// Every percentile from one sort.
  ///
  /// Reading p50, p95 and p99 as three separate calls sorts the ring three times, and the
  /// reporter reads seven rings every second. At 30 fps that is tens of thousands of needless
  /// comparisons per second on the same queue that packs frames — the measurement showing up
  /// inside the thing being measured.
  public func distribution() -> PreviewDistribution {
    guard count > 0 else { return .empty }
    let sorted = samples.prefix(count).sorted()
    func at(_ quantile: Double) -> Double {
      let position = Int((Double(sorted.count - 1) * quantile).rounded())
      return Double(sorted[min(max(position, 0), sorted.count - 1)]) / 1_000_000.0
    }
    return PreviewDistribution(p50: at(0.5), p95: at(0.95), p99: at(0.99), max: maxMs, count: count)
  }

  public func percentileMs(_ quantile: Double) -> Double {
    guard count > 0 else { return 0 }
    let sorted = samples.prefix(count).sorted()
    let position = Int((Double(sorted.count - 1) * quantile).rounded())
    let clamped = min(max(position, 0), sorted.count - 1)
    return Double(sorted[clamped]) / 1_000_000.0
  }

  public func reset() {
    count = 0
    index = 0
    maxNs = 0
  }
}

/// Counters for one preview subscription.
///
/// Every drop has its own counter on purpose: "delivered fps is low" is not a finding, but
/// "delivered is low because the pacer skipped, not because the consumer was busy" is.
///
/// The timing rings are split finer than they strictly need to be for the same reason. A single
/// `sendMs` would average an enqueue that returns immediately together with a copy that has not
/// happened yet, and the two platforms would then be reporting different things under one name.
public final class PreviewStats {
  public private(set) var sourceFrames: Int = 0
  public private(set) var admitted: Int = 0
  public private(set) var skippedPacing: Int = 0
  public private(set) var skippedBusy: Int = 0
  public private(set) var delivered: Int = 0
  public private(set) var payloadBytes: Int = 0
  public private(set) var ackTimeouts: Int = 0
  public private(set) var staleAcks: Int = 0
  public private(set) var unsupportedFormat: Int = 0
  public private(set) var transportErrors: Int = 0
  public private(set) var packFailures: Int = 0
  /// Frames refused on the decoder thread because the worker had not finished the previous one.
  /// Counted separately from `skippedBusy` so a decoder-side refusal is not confused with a
  /// consumer-side one; they have different fixes.
  public private(set) var preDispatchDrops: Int = 0
  /// The pack worker wanted a send buffer and both were still held by the transport.
  public private(set) var slotStarved: Int = 0

  /// Producing the source frame. Only the synthetic source pays this; a real decoder hands over
  /// a buffer it has already made. Kept apart from `pack` so a slow test pattern cannot be read
  /// as a slow pipeline.
  public let generate = PreviewPercentileRing()
  public let pack = PreviewPercentileRing()
  /// Time to hand the frame to the transport. On iOS this returns before the copy completes.
  public let sendEnqueue = PreviewPercentileRing()
  /// Time until the transport reports it is finished with our buffer. The honest send cost.
  public let sendComplete = PreviewPercentileRing()
  public let roundTrip = PreviewPercentileRing()
  /// Admission to the first instruction of the pack worker: the handoff cost, not the work.
  public let admitToPack = PreviewPercentileRing()
  /// Gap between consecutive delivered frames. Mean fps hides a 30/100 ms alternation; this does not.
  public let deliveryGap = PreviewPercentileRing()

  /// Start of run to first delivered frame, and to its acknowledgement. A slow cold start is
  /// invisible in steady-state fps but very visible to someone opening the panel.
  public private(set) var firstDeliveredLatencyMs: Double = 0
  public private(set) var firstAckLatencyMs: Double = 0

  // Optionals rather than a zero sentinel. A monotonic clock reading of exactly zero is unlikely
  // on a device but trivial in a test, and "not yet set" and "set to zero" have to stay distinct
  // or the first gap of every run goes unrecorded.
  private var runStartNs: Int64?
  private var lastDeliveredAtNs: Int64?
  private var hasFirstDelivered = false
  private var hasFirstAck = false

  /// Counters since the last `takeWindow`, so a 1 Hz report can show rates without the caller
  /// keeping its own deltas.
  private var windowSourceFrames = 0
  private var windowDelivered = 0
  private var windowPayloadBytes = 0
  private var windowStartNs: Int64?

  public init() {}

  public func onSourceFrame() { sourceFrames += 1; windowSourceFrames += 1 }
  public func onAdmitted() { admitted += 1 }
  public func onSkippedPacing() { skippedPacing += 1 }
  public func onSkippedBusy() { skippedBusy += 1 }
  public func onAckTimeout() { ackTimeouts += 1 }
  public func onStaleAck() { staleAcks += 1 }
  public func onUnsupportedFormat() { unsupportedFormat += 1 }
  public func onTransportError() { transportErrors += 1 }
  public func onPackFailure() { packFailures += 1 }
  public func onPreDispatchDrop() { preDispatchDrops += 1 }
  public func onSlotStarved() { slotStarved += 1 }

  /// Mark the beginning of a run so the first-frame latencies have an origin.
  ///
  /// This also restarts the rate window. Without that, the first second after a stop/start would
  /// be averaged over however long the preview sat idle and report a frame rate near zero.
  public func onRunStart(nowNs: Int64) {
    runStartNs = nowNs
    lastDeliveredAtNs = nil
    hasFirstDelivered = false
    hasFirstAck = false
    firstDeliveredLatencyMs = 0
    firstAckLatencyMs = 0
    windowStartNs = nowNs
    windowSourceFrames = 0
    windowDelivered = 0
    windowPayloadBytes = 0
  }

  public func onDelivered(bytes: Int, nowNs: Int64) {
    delivered += 1
    windowDelivered += 1
    payloadBytes += bytes
    windowPayloadBytes += bytes
    if let last = lastDeliveredAtNs {
      deliveryGap.record(nowNs - last)
    }
    lastDeliveredAtNs = nowNs
    if !hasFirstDelivered, let start = runStartNs {
      hasFirstDelivered = true
      firstDeliveredLatencyMs = Double(nowNs - start) / 1_000_000.0
    }
  }

  public func onAckAccepted(roundTripNs: Int64, nowNs: Int64) {
    roundTrip.record(roundTripNs)
    if !hasFirstAck, let start = runStartNs {
      hasFirstAck = true
      firstAckLatencyMs = Double(nowNs - start) / 1_000_000.0
    }
  }

  public struct Window {
    public let sourceFps: Double
    public let deliveredFps: Double
    public let bytesPerSecond: Double
  }

  /// Read and restart the rate window. Returns zeros for the first call of a subscription.
  public func takeWindow(nowNs: Int64) -> Window {
    defer {
      windowStartNs = nowNs
      windowSourceFrames = 0
      windowDelivered = 0
      windowPayloadBytes = 0
    }
    guard let start = windowStartNs, nowNs > start else {
      return Window(sourceFps: 0, deliveredFps: 0, bytesPerSecond: 0)
    }
    let seconds = Double(nowNs - start) / 1_000_000_000.0
    return Window(
      sourceFps: Double(windowSourceFrames) / seconds,
      deliveredFps: Double(windowDelivered) / seconds,
      bytesPerSecond: Double(windowPayloadBytes) / seconds
    )
  }

  public func reset(nowNs: Int64) {
    sourceFrames = 0
    admitted = 0
    skippedPacing = 0
    skippedBusy = 0
    delivered = 0
    payloadBytes = 0
    ackTimeouts = 0
    staleAcks = 0
    unsupportedFormat = 0
    transportErrors = 0
    packFailures = 0
    preDispatchDrops = 0
    slotStarved = 0
    generate.reset()
    pack.reset()
    sendEnqueue.reset()
    sendComplete.reset()
    roundTrip.reset()
    admitToPack.reset()
    deliveryGap.reset()
    firstDeliveredLatencyMs = 0
    firstAckLatencyMs = 0
    hasFirstDelivered = false
    hasFirstAck = false
    runStartNs = nowNs
    lastDeliveredAtNs = nil
    windowStartNs = nowNs
    windowSourceFrames = 0
    windowDelivered = 0
    windowPayloadBytes = 0
  }
}
